File: articles/calculus-for-systems-modeling/articles/directional-derivatives-and-gradients/fortran/directional_derivative_audit.f90
program directional_derivative_audit
  implicit none
  real(8), dimension(3) :: xs, ys, vxs, vys, steps
  real(8) :: x, y, vx, vy, ux, uy, norm_value, deriv, estimated, actual, error_value
  integer :: i
  logical :: feasible

  xs = (/4.0d0, 4.0d0, 8.0d0/)
  ys = (/3.0d0, 3.0d0, 1.0d0/)
  vxs = (/1.0d0, 2.0d0, 1.0d0/)
  vys = (/1.0d0, -1.0d0, 1.0d0/)
  steps = (/0.25d0, 0.25d0, 1.0d0/)

  print '(A)', 'x y vx vy ux uy gradient_x gradient_y derivative step estimated actual error feasible'

  do i=1,3
    x = xs(i)
    y = ys(i)
    vx = vxs(i)
    vy = vys(i)
    norm_value = sqrt(vx*vx + vy*vy)
    ux = vx / norm_value
    uy = vy / norm_value
    deriv = gx(x,y)*ux + gy(x,y)*uy
    estimated = steps(i) * deriv
    actual = f(x+steps(i)*ux, y+steps(i)*uy) - f(x,y)
    error_value = abs(actual - estimated)
    feasible = x >= 0.0d0 .and. y >= 0.0d0 .and. x + y <= 10.0d0 .and. x + steps(i)*ux >= 0.0d0 .and. y + steps(i)*uy >= 0.0d0 .and. x + steps(i)*ux + y + steps(i)*uy <= 10.0d0
    if (feasible) then
      print '(14F12.6,1X,A)', x, y, vx, vy, ux, uy, gx(x,y), gy(x,y), deriv, steps(i), estimated, actual, error_value, 1.0d0, 'feasible'
    else
      print '(14F12.6,1X,A)', x, y, vx, vy, ux, uy, gx(x,y), gy(x,y), deriv, steps(i), estimated, actual, error_value, 0.0d0, 'infeasible'
    end if
  end do

contains
  function f(x,y) result(value)
    real(8), intent(in) :: x, y
    real(8) :: value
    value = 3.0d0*x + 2.0d0*y + 0.5d0*x*y
  end function f

  function gx(x,y) result(value)
    real(8), intent(in) :: x, y
    real(8) :: value
    value = 3.0d0 + 0.5d0*y
  end function gx

  function gy(x,y) result(value)
    real(8), intent(in) :: x, y
    real(8) :: value
    value = 2.0d0 + 0.5d0*x
  end function gy
end program

File: articles/calculus-for-systems-modeling/articles/directional-derivatives-and-gradients/cpp/directional_derivative_audit.cpp
#include <cmath>
#include <iomanip>
#include <iostream>
#include <stdexcept>
#include <vector>

double f(double x, double y){ return 3.0*x + 2.0*y + 0.5*x*y; }
double gx(double, double y){ return 3.0 + 0.5*y; }
double gy(double x, double){ return 2.0 + 0.5*x; }
std::pair<double,double> normalize(double vx, double vy){ double norm = std::sqrt(vx*vx + vy*vy); if(norm == 0.0) throw std::runtime_error("Direction vector must be nonzero."); return {vx/norm, vy/norm}; }
double directional_derivative(double x, double y, double ux, double uy){ return gx(x,y)*ux + gy(x,y)*uy; }
bool feasible_direction(double x, double y, double ux, double uy, double step){ return x >= 0.0 && y >= 0.0 && x+y <= 10.0 && x+step*ux >= 0.0 && y+step*uy >= 0.0 && x+step*ux+y+step*uy <= 10.0; }

int main(){
  std::vector<std::vector<double>> cases = {{4.0,3.0,1.0,1.0,0.25},{4.0,3.0,2.0,-1.0,0.25},{8.0,1.0,1.0,1.0,1.0}};
  std::cout << std::fixed << std::setprecision(12);
  std::cout << "x,y,direction_x,direction_y,unit_x,unit_y,gradient_x,gradient_y,directional_derivative,step_size,estimated_change,actual_change,absolute_error,feasible_direction,warning\n";
  for(auto row : cases){
    double x=row[0], y=row[1], vx=row[2], vy=row[3], step=row[4];
    auto unit = normalize(vx,vy);
    double ux=unit.first, uy=unit.second;
    double deriv=directional_derivative(x,y,ux,uy);
    double estimated=step*deriv;
    double actual=f(x+step*ux,y+step*uy)-f(x,y);
    bool feasible=feasible_direction(x,y,ux,uy,step);
    std::cout << x << "," << y << "," << vx << "," << vy << "," << ux << "," << uy << "," << gx(x,y) << "," << gy(x,y) << "," << deriv << "," << step << "," << estimated << "," << actual << "," << std::abs(actual-estimated) << "," << feasible << "," << (feasible ? "" : "Direction and step move outside the feasible region.") << "\n";
  }
}
